pkg/gardenlet/bootstrap/certificate/certificate_rotation_test.go

```diff
@@ -0,0 +1,342 @@
+	Describe("#rotateCertificate", func() {
+		It("should not return an error", func() {
+			kubeClient := fake.NewSimpleClientset(&approvedCSR)
+			k8sGardenClient.EXPECT().Kubernetes().Return(kubeClient)
+
+			// mock the update of the secret in the seed with the rotated kubeconfig
+			certClientConfig := &rest.Config{Host: "testhost", TLSClientConfig: rest.TLSClientConfig{
+				Insecure: false,
+				CAFile:   "filepath",
+			}}
+
+			k8sGardenClient.EXPECT().RESTConfig().Return(certClientConfig)
+			k8sGardenClient.EXPECT().DirectClient().Return(k8sClient)
+
+			fakeClientMap := fakeclientmap.NewClientMap().
+				AddClient(keys.ForGarden(), k8sGardenClient).
+				AddClient(keys.ForSeed(seed), k8sGardenClient)
+
+			err := rotateCertificate(context.TODO(), log, gardenClientConnection, fakeClientMap, seedName)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		It("should return an error - CSR is denied", func() {
+			kubeClient := fake.NewSimpleClientset(&deniedCSR)
+			k8sGardenClient := fakeclientset.NewClientSetBuilder().WithKubernetes(kubeClient).Build()
+			fakeClientMap := fakeclientmap.NewClientMap().
+				AddClient(keys.ForGarden(), k8sGardenClient).
+				AddClient(keys.ForSeed(seed), k8sGardenClient)
+
+			err := rotateCertificate(context.TODO(), log, gardenClientConnection, fakeClientMap, seedName)
+			Expect(err).To(HaveOccurred())
+		})
+	})
+
+	Describe("Tests that require a generated kubeconfig with a client certificate", func() {
+		var (
+			gardenKubeconfigWithValidClientCert   string
+			gardenKubeconfigWithInValidClientCert string
+			baseKubeconfig                        = `
+apiVersion: v1
+clusters:
+- cluster:
+    server: https://localhost:2443
+  name: gardenlet
+contexts:
+- context:
+    cluster: gardenlet
+    user: gardenlet
+  name: gardenlet
+current-context: gardenlet
+kind: Config
+preferences: {}
+users:
+- name: gardenlet
+  user:
+    client-certificate-data: %s
+    client-key-data: %s
+`
+		)
+
+		Describe("#getCurrentCertificate", func() {
+			BeforeEach(func() {
+				k8sGardenClient.EXPECT().DirectClient().Return(k8sClient)
+				validity := 20 * time.Second
+				cert := generateCertificate(validity)
+				gardenKubeconfigWithValidClientCert = fmt.Sprintf(baseKubeconfig, utils.EncodeBase64(cert.CertificatePEM), utils.EncodeBase64(cert.PrivateKeyPEM))
+				gardenKubeconfigWithInValidClientCert = fmt.Sprintf(baseKubeconfig, "bm90LXZhbGlk", utils.EncodeBase64(cert.PrivateKeyPEM))
+			})
+
+			It("should not return an error", func() {
+				// mock existing secret with garden kubeconfig
+				updateKubeconfigSecret(k8sClient, gardenClientConnection, gardenKubeconfigWithValidClientCert)
+
+				fakeClientMap := fakeclientmap.NewClientMap().
+					AddClient(keys.ForGarden(), k8sGardenClient).
+					AddClient(keys.ForSeed(seed), k8sGardenClient)
+
+				cert, err := getCurrentCertificate(context.TODO(), gardenClientConnection, fakeClientMap, seedName)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(cert).ToNot(BeNil())
```

Comment:
Can you check, that `cert` has the right content? Otherwise, this test is kind of inaccurate.

Reply:
I think what is being tested is enough.  Otherwise the whole TLS cert parsing procedure has to be done again, only to check if the cert really has the correct common name and org name.
I do not see the point in checking those fields if the test itself mocks the certificate in the CSR.

If the function under test suceeds that means
 - that there was indeed a cert in the CSR status
- that cert is a valid x509 cert and is properly formatted.